src/function/vector_cast_functions.cpp

```diff
@@ -104,12 +104,73 @@ static void nestedTypesCastExecFunction(const std::vector<std::shared_ptr<ValueV
         reinterpret_cast<CastFunctionBindData*>(dataPtr));
 }
 
-bool CastFunction::hasImplicitCast(const LogicalType& srcType, const LogicalType& dstType) {
-    // TODO(Jiamin): should remove after support list implicit cast
-    if (srcType.getLogicalTypeID() == LogicalTypeID::LIST &&
-        dstType.getLogicalTypeID() == LogicalTypeID::LIST) {
+bool CastFunction::hasImplicitCastList(const LogicalType& srcType, const LogicalType& dstType) {
+    return hasImplicitCast(*ListType::getChildType(&srcType), *ListType::getChildType(&dstType));
+}
+
+bool CastFunction::hasImplicitCastArray(const LogicalType& srcType, const LogicalType& dstType) {
+    if (ArrayType::getNumElements(&srcType) != ArrayType::getNumElements(&dstType)) {
         return false;
     }
+    return hasImplicitCast(*ArrayType::getChildType(&srcType), *ArrayType::getChildType(&dstType));
+}
+
+bool CastFunction::hasImplicitCastStruct(const LogicalType& srcType, const LogicalType& dstType) {
+    auto srcFields = StructType::getFields(&srcType), dstFields = StructType::getFields(&dstType);
+    if (srcFields.size() != dstFields.size()) {
+        return false;
+    }
+    for (auto i = 0u; i < srcFields.size(); i++) {
+        if (srcFields[i]->getName() != dstFields[i]->getName()) {
+            return false;
+        }
+        if (!hasImplicitCast(*srcFields[i]->getType(), *dstFields[i]->getType())) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool CastFunction::hasImplicitCastUnion(const LogicalType& /*srcType*/,
+    const LogicalType& /*dstType*/) {
+    // todo: implement union casting function
+    // currently, there seems to be no casting functionality between union types
+    return false;
+}
+
+bool CastFunction::hasImplicitCastMap(const LogicalType& srcType, const LogicalType& dstType) {
+    auto srcKeyType = *MapType::getKeyType(&srcType);
+    auto srcValueType = *MapType::getValueType(&srcType);
+    auto dstKeyType = *MapType::getKeyType(&dstType);
+    auto dstValueType = *MapType::getValueType(&dstType);
+    return hasImplicitCast(srcKeyType, dstKeyType) && hasImplicitCast(srcValueType, dstValueType);
+}
+
+bool CastFunction::hasImplicitCast(const LogicalType& srcType, const LogicalType& dstType) {
+    if ((LogicalTypeUtils::isNested(srcType) &&
```

Comment:
Can't we just reuse `tryGetMaxLogicalType`?

Reply:
I wouldn't do that here; max logical type is defined for combining two separate types and doesn't exactly follow implicit casting rules.